adam/perception/perception_graph.py

```diff
@@ -2932,9 +2876,225 @@ def _translate_frames(
             (len(perceptual_representation.frames),),
         )
 
-        raise NotImplementedError(
-            "Multiple frame translation is not yet supported for Visual Perception Frames."
+        if perceptual_representation.simulated_actions_features is None:
+            raise ValueError(
+                "Simulated actions features cannot be None when creating a multi-frame perception graph."
+            )
+
+        _dynamic_digraph = _merge_before_and_after_graphs(
+            self._translate_frame(perceptual_representation.frames[0]),
+            self._translate_frame(perceptual_representation.frames[1]),
+        )
+
+        decoded_object_to_cluster_node = {
+            object_cluster_node.cluster_id: object_cluster_node
+            for object_cluster_node in _dynamic_digraph.nodes
+            if isinstance(object_cluster_node, ObjectClusterNode)
+        }
+
+        action_stroke_graph = perceptual_representation.simulated_actions_features[
+            "stroke_graph"
+        ]
+        agent_node = decoded_object_to_cluster_node[
+            perceptual_representation.simulated_actions_features["object_name"]
+        ]
+        involved_object_nodes = [
+            decoded_object_to_cluster_node[object_name]
+            for object_name in perceptual_representation.simulated_actions_features[
+                "objects"
+            ]
+        ]
+        involved_object_nodes.append(agent_node)
+
+        action_decode_node = CategoricalNode(
+            label="action_gnn",
+            value=action_stroke_graph["action_name"].lower(),
+            weight=action_stroke_graph["confidence_score"],
+        )
+
+        joint_point_nodes = [
+            JointPointNode(
+                scene_xyd_coord=DepthPoint(xyd_coord[0], xyd_coord[1], xyd_coord[2]),
+                world_coord=Point(world_coord[0], world_coord[1], world_coord[2]),
+                confidence=conf,
+                weight=conf,
+                temporal_index=i,
+            )
+            for i, (world_coord, xyd_coord, conf) in enumerate(
+                zip(
+                    action_stroke_graph["joint_points_world_coords"],
+                    action_stroke_graph["joint_points_xyd_coords"],
+                    action_stroke_graph["joint_points_confidence_score"],
+                )
+            )
+        ]
```

Comment:
Hmm, this code expects to get a sequence of 3 floats, a sequence of 3 floats, and a single float out of iterating over this zip, but when I try that locally with the `walk.yaml` data (not running all the surrounding code, but just loading the action features and stroke graph manually in the interpreter) I'm getting *lists of lists*.

Thinking of these as arrays, the xyd and world coords for the joint points should be of shape T x K x 3. When we zip them we're getting (I think) K x 3 slices. (For confidence we're getting a K-size slice/list.) I think what we want here is another zip:
```suggestion
        joint_point_nodes = [
            JointPointNode(
                scene_xyd_coord=DepthPoint(xyd_coord[0], xyd_coord[1], xyd_coord[2]),
                world_coord=Point(world_coord[0], world_coord[1], world_coord[2]),
                confidence=conf,
                weight=conf,
                temporal_index=i,
            )
            for i, (world_coords, xyd_coords, confs) in enumerate(
                zip(
                    action_stroke_graph["joint_points_world_coords"],
                    action_stroke_graph["joint_points_xyd_coords"],
                    action_stroke_graph["joint_points_confidence_score"],
                )
            )
            for world_coord, xyd_coord, conf in zip(
                world_coords, xyd_coords, confs
            )
        ]
```

Reply:
🤔  That's right there are K joint points we are tracking so we need to iterate over those too. I may be getting close to agreeing with you that this should maybe be a static function even though its just a generator... this is a very complicated generator. I'm thinking of doing the following and tracking a `node_index` as well as a `temporal_index`. This means I should probably check with ASU that the joints are always in the same order between decode outputs.

```py
joint_point_nodes = [
    JointPointNode(
        scene_xyd_coord=DepthPoint(xyd_coord[0], xyd_coord[1], xyd_coord[2]),
        world_coord=Point(world_coord[0], world_coord[1], world_coord[2]),
        confidence=conf,
        weight=conf,
        temporal_index=temporal_index,
        joint_index=joint_index,
    )
    for joint_index, (world_coords, xyd_coords, confs) in enumerate(
        zip(
            action_stroke_graph["joint_points_world_coords"],
            action_stroke_graph["joint_points_xyd_coords"],
            action_stroke_graph["joint_points_confidence_score"],
        )
    )
    for temporal_index, (world_coord, xyd_coord, conf) in enumerate(
        zip(
            world_coords, xyd_coords, conf
        )
    )
]
```

Adjacent to this topic, how should i go about matching these nodes?

Reply:
The single-loop generator looked okay -- I agree though, with the two loops it's looking pretty complicated and probably worth a staticmethod. That code makes sense to me -- it probably is worth tracking the joint index as well for debugging purposes.

Yes, I think that's a good idea to check with ASU about joint ordering. Another question to check is, is the time dimension first or the joint dimension? i.e. is it T x K x 3 or K x T x 3. The above code would use the latter -- I thought time came first but I'm not confident in that so probably worth checking.

Reply:
One could even go further and rewrite the method as building up a list using loops -- more like a generator function -- which makes the control flow slightly more straightforward (top to bottom) but I'm not going to request that. Thanks for extracting it.